api/callbacks.go

```diff
@@ -89,29 +128,41 @@ func buildAPI(api *GoAPI) C.GoApi {
 }
 
 //export cHumanAddress
-func cHumanAddress(ptr *C.api_t, canon C.Buffer, human C.Buffer) i32 {
+func cHumanAddress(ptr *C.api_t, canon C.Buffer, human *C.Buffer) (ret C.GoResult) {
+	defer recoverPanic(&ret)
+	if human == nil {
+		// we received an invalid pointer
+		return C.GoResult_BadArgument
+	}
+
 	api := (*GoAPI)(unsafe.Pointer(ptr))
 	c := receiveSlice(canon)
 	h, err := api.HumanAddress(c)
 	if err != nil {
-		return -1
-	}
-	if len(h) == 0 {
-		return 0
+		return C.GoResult_Other
 	}
-	return i32(writeToBuffer(human, []byte(h)))
+	*human = allocateRust([]byte(h))
+	return C.GoResult_Ok
 }
 
 //export cCanonicalAddress
-func cCanonicalAddress(ptr *C.api_t, human C.Buffer, canon C.Buffer) i32 {
+func cCanonicalAddress(ptr *C.api_t, human C.Buffer, canon *C.Buffer) (ret C.GoResult) {
+	defer recoverPanic(&ret)
+	if canon == nil {
+		// we received an invalid pointer
+		return C.GoResult_BadArgument
+	}
+
 	api := (*GoAPI)(unsafe.Pointer(ptr))
 	h := string(receiveSlice(human))
 	c, err := api.CanonicalAddress(h)
 	if err != nil {
-		return -1
+		return C.GoResult_Other
 	}
-	if len(c) == 0 {
-		return 0
+	if c != nil {
+		*canon = allocateRust(c)
 	}
-	return i32(writeToBuffer(canon, c))
+
+	// If we do not set canon to a meaningful value, then the other side will interpret that as an empty result.
```

Comment:
Yeah, we can execute a Go `panic()` here which would return `C.GoResult_Panic`, or return `C.GoResult_Other`. What do you think is more appropriate?

Reply:
Oh, the same also applies to cases where Humanize or Canonicalize return an empty string/buffer. The code just returns those as they are at the moment, but we can check for that as well here.

Reply:
A go `panic("this must not happen")` is perfect. No need to propagate it somewhere, since this is clearly a programming error, no a case that can happen depending on user input.

Reply:
Alright :+1: 
Just keep in mind if it does happen, it will be reported back to rust as `C.GoResult_Panic`.
Generally i imagine that panics will be thrown if a bug is triggered. Should we maybe add a log of all such panics in `recoverPanic`? It'll help in debugging if it does happen.

Reply:
Sounds plausible. I don't have any strong opinion other than _crash as hard as possible_ to ensure such unexpected states are fund during development.

Reply:
As long as later PRs correctly treat the `GoResult::Panic` case, and effectively eventually crash the node, or at least fail the tx with a big red alert, that's as close as you can get :) at the moment these variants are not treated correctly in the rust side, yet.

Reply:
This is implemented in 46857e7 and 09cf731